### ISSUE TITLE
Codecs: add support for opus

### ISSUE THREAD

**Issue by tiagogaspar8** [peak3d/inputstream.adaptive#885]
Hi!

I wanted to ask for support for the opus audio codec instead of the current aac.
Many add-ons could use it but since inputstream adaptive doesn't support it it can't be used.

Thanks!

**Comment by glennguy** [comment]
Hi Tiago,

> Many add-ons could use it

Do you have any examples of streams that use opus? I haven't seen it myself.

**Comment by anxdpanic** [comment]
I have it disabled in the YouTube add-on, a user just reported it's working now. So I will have to investigate further when I get a chance.

https://github.com/anxdpanic/plugin.video.youtube/blob/master/resources/lib/youtube_plugin/kodion/impl/xbmc/xbmc_context.py#L315

**Comment by tiagogaspar8** [comment]
Well, for now I know that YouTube uses it, and I believe I have seen twitch use it (?), Not sure, but they could right? I believe most services that use vp9 tend to pair it with opus.

**Comment by glennguy** [comment]
Cool, thanks for the info - I'm probably way out of my depth with this one but I will look into it once refactoring/cleanup is done.

**Comment by tiagogaspar8** [comment]
Glad to hear it, if you need any help in testing just say so!

**Comment by CastagnaIT** [comment]
i close this issue since OPUS codec is already working
i found a sample stream where test it

**Comment by tiagogaspar8** [comment]
Hi @CastagnaIT great news then!

I have a question that you can maybe help me with. Recently _some_ add-ons that use inputstream adaptive have been plagued with an issue where they are "rate limited" out of the blue. Now, my question to you is, can this maybe be related to inputstream or a change to kodi itself?

Thanks!😁

**Comment by CastagnaIT** [comment]
explain what means: "rate limited" out of the blue

**Comment by tiagogaspar8** [comment]
I've tested this in two add-ons, the main one is YouTube.
When some streams are played, they are rate limited, this is clearly noticed while opening task manager and watching the network speed.
For example, video A, when it starts playing, shows a spike in network activity at the level of 300+ Mbps, it loads well and plays well too. Video B, when it starts playing shows network traffic reaching 50-60mbps and the video plays for 3 seconds, then there is a lot of lag, like it's not even realising it needs to buffer. All videos are 4k.
On YouTube, there isn't a type of video that triggers it specifically, but in general lines 4k SDR videos, OK (some exceptions apply), and 4k HDR videos trigger this issue (I haven't seen one that doesn't).
The other add-on that faces this issue is called RTPlay and all videos face this issue (and they aren't HDR) but I think it can't be used outside Portugal.

**Comment by CastagnaIT** [comment]
> opening task manager and watching the network speed

this is not reliable way to check whats wrong
because in this way you dont take in account of the software buffering which is deferred

i cannot give a precise answer due to lack of info and logs, and i don't know which Kodi version you mean
now we focus on Kodi 20 only, for Kodi 19 we backport only critical fixes
ISA for Kodi 20 is different than 19 because have a double buffering (kodi/isa)

> On YouTube, there isn't a type of video that triggers it specifically, but in general lines 4k SDR videos, OK (some exceptions apply), and 4k HDR videos trigger this issue

also here is so difficult give an answer, also HDR things are still in WIP on Kodi 20

There is a working youtube addon in kodi? I have tried both Youtube and Tubed addon both raise same api error also using google API dev

EDIT: can you provide addon github for RTPlay ?

**Comment by tiagogaspar8** [comment]
> this is not reliable way to check whats wrong
because in this way you dont take in account of the software buffering which is deferred

That's the best way I knew how 🥲 
I don't even know what you mean by software buffering 😢 , what I can tell you is that it doesn't have to do with connectivity, the network is well and fast

> i cannot give a precise answer due to lack of info and logs, and i don't know which Kodi version you mean
now we focus on Kodi 20 only, for Kodi 19 we backport only critical fixes
ISA for Kodi 20 is different than 19 because have a double buffering (kodi/isa)

I'm currently using the master builds

> also here is so difficult give an answer, also HDR things are still in WIP on Kodi 20

HDR is just a way to profile the general type of videos that are broken

> There is a working youtube addon in kodi? I have tried both Youtube and Tubed addon both raise same api error also using google API dev

Sure! The add-on called Youtube works if you only use it to play videos (no likes or other features), you can use the mobile app and share videos to the app to be sent to kodi or use like I do an extension for chrome called Play to Kodi and it will work beautifully.

> EDIT: can you provide addon github for RTPlay ?

Sure can! I can also provide the logs! I was just asking if it's worth creating a new issue here or in the main repo just to separate stuff 😄 

https://github.com/enen92/plugin.video.rtpplay

The HDR one is the broken one with the video and the nohdr one is the one that works well.
[hdr.log](https://github.com/xbmc/inputstream.adaptive/files/8607755/hdr.log)
[nohdr.log](https://github.com/xbmc/inputstream.adaptive/files/8607756/nohdr.log)

BTW, thanks for the help and attention.

**Comment by CastagnaIT** [comment]
for now i tested RTP Play and do not use InputStreamAdaptive but ffmpeg then is not here the problem

**Comment by CastagnaIT** [comment]
i tested youtube with a 4k 60fps by enabling DASH+Adaptive VP9 setting
and used "ask quality" in ISA stream selection setting

if i play the 4k video on webbrowser:
the 4k is played not so fluid and often stop for buffering (i have ~50Mbit network but it should fit)

if i play on kodi, i think results are similar to your test:
play first seconds, but not so fluid, then freeze (needs buffering) my CPU starts to be medium/highly busy, after go on etc..

there are multiple factors that influence:
1) 4k stream has a very high bitrate ~30Mbit in my test case
2) software video decoding, at least with my video card on Windows (depends also by hardware used)
3) the buffering management in ISA

hardware decoding is important to fluidity, and to leave CPU free, in my case seem i cannot have it,
the buffering management in ISA could affect this problem more
but we already know that buffer management in ISA need to be optimized
it is a part that will be done in the future because need a full rework and study how can be done in better way

there are currently no alternative workaround to this problem
we will consider this use case for future tests

**Comment by tiagogaspar8** [comment]
Well, I can tell you that in January for sure this issue wasn't happening, so it was probably a change made since then right?

Do you have any suggestion for troubleshooting? To figure out what's wrong?

I can tell you that my laptop has hardware acceleration and that cpu usage doesn't climb above 20%, either on a browser or on kodi, the only noticeable thing really is the traffic shape in task manager.

**Comment by CastagnaIT** [comment]
the only way to help us is that you find the last working Nexus version
you should build the addon yourself for each version to test (there may however be compatibility issues to be resolved)
this can be a first start, that however not guaranteed nothing because we have done and we are making major changes to the code

**Comment by tiagogaspar8** [comment]
Hi,
Sorry for the big delay in answering.
Unfortunately, I believe I can't do that, not only because I don't have the time but because I have no idea if I can do it.
What I can do is create an issue about it with the details and report back every time I upgrade.
My question is: should I create it here or in the main repo? I believe it might have to do with inputstream but it might have to do with the main repo, which is weird since I don't believe it is affected by it.

**Comment by CastagnaIT** [comment]
open a new issue would be useful just in case you know the latest working version otherwise there is no point

this because we are already aware of the improvements to be made in the buffering management of the addon which will most likely solve this use case as well

this is already planned
but it will take a long time to be completed

**Comment by tiagogaspar8** [comment]
Ok, I have some bad news about this...
I use Kodi to manage media that I access through an smb share and I configured Kodi to access the files via the mount point, and I just noticed that the issue also appears, but in this case not with buffering but when trying to browse the timeline it seems to be stuck at 40mbps and this isn't using Kodi's builtin smb browser but the window's one.

This means that it has nothing to do with inputstream right? this is Kodi itself I believe.

Again, sorry to bother, just thought it was best to inform of all the data I can find.
